### PR TITLE
Fix handling of v2 hashes

### PR DESCRIPTION
**Pull request by bernt-matthias**
Added tests first for the mulled container resolvers here https://github.com/galaxyproject/galaxy/pull/18522/commits/5b3c4bddd38d88bd195176be9f20f3e81d5336df

Then fixed (and more tests) here: https://github.com/galaxyproject/galaxy/pull/18522/commits/e7bca837f08345980f240f633921dbcce174345e

See details in https://github.com/galaxyproject/galaxy/pull/18522/commits/e7bca837f08345980f240f633921dbcce174345e

I'm tempted to backport this to 24.1 in order to make this effective in the multi-package-container repo.

## How to test the changes?
(Select all options that apply)
- [x] I've included appropriate [automated tests](https://docs.galaxyproject.org/en/latest/dev/writing_tests.html).
- [ ] This is a refactoring of components with existing test coverage.
- [ ] Instructions for manual testing are as follows:
  1. [add testing steps and prerequisites here if you didn't write automated tests covering all your changes]

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

### CODE REVIEW

File: test/functional/tools/mulled_example_multi_2.xml [hunk 1]
```diff
@@ -0,0 +1,79 @@
+<!-- same as mulled_example_multi_1, but requirements use builds
+     used to show that this does not yield the same mulled v2 hash,
+     i.e. the build is considered as part of version -->
+<tool id="mulled_example_multi_2" name="mulled_example_multi_2" version="0.1.0" license="AFL-3.0">
```
Comment (mvdbeek) [inline test/functional/tools/mulled_example_multi_2.xml]: The unit tests cover this already, I would prefer that we don't add this. In the end this is just a bug in mulled_build_tool.py which we don't use in Galaxy.

**Review by mvdbeek — APPROVED** [review]

File: test/functional/tools/mulled_example_multi_2.xml (hunk 1 above)
Comment (bernt-matthias) [reply test/functional/tools/mulled_example_multi_2.xml]: > The unit tests cover this already, I would prefer that we don't add this. In the end this is just a bug in mulled_build_tool.py which we don't use in Galaxy.

I can remove it. The idea was to make sure that the parser really gives a version string with the build (maybe this test should be added somewhere) .. but we would realize very quickly if something goes wrong here anyway.

**Review by mvdbeek — CHANGES_REQUESTED** [review]
Wait, this can't be right, why did you remove the build parsing ?

**Comment by bernt-matthias** [comment]
I only remove it from the function that parses targets from a string (comma separated list of tools and versions) which is used in `build-hash` and the other two tools.

The remaining code (e.g. the mulled container resolver etc)  keeps the build as part of the version. I think that the tests which I added in my first commit show this (the version part of the hash differs when builds are added, i.e. the build is considered as part of the version when computing the hash). 

Wondering if this got a bit confusing when we merged `Target` and `CondaTarget` https://github.com/galaxyproject/galaxy/commit/d65721c78de7adf9ad749a0e07d0dc7ad8684c2a (guess for conda we parse the build).

File: lib/galaxy/tool_util/deps/mulled/mulled_build.py [hunk 2]
```diff
@@ -531,12 +531,7 @@ def target_str_to_targets(targets_raw):
     def parse_target(target_str):
         if "=" in target_str:
             package_name, version = target_str.split("=", 1)
-            build = None
-            if "=" in version:
-                version, build = version.split("=")
-            elif "--" in version:
-                version, build = version.split("--")
-            target = build_target(package_name, version, build)
+            target = build_target(package_name, version)
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/deps/mulled/mulled_build.py]: I can't see how that won't break on:
```
combinations/hash.tsv
79:bedtools=2.17.0--0,python=2.7.13	bgruening/busybox-bash:0.1	0
94:picard=2.22.8--0,bwa=0.7.17	bgruening/busybox-bash:0.1	0
95:whatshap=1.0--py37h9a982cc_1,tabix=0.2.6--ha92aebf_0	bgruening/busybox-bash:0.1	0
```
Was there any reason not to use my suggestion ?

Comment (bernt-matthias) [reply lib/galaxy/tool_util/deps/mulled/mulled_build.py]: Indeed we would change image names, but the new image names would be the ones which the container resolvers are looking for (in the current implementation).

Comment (mvdbeek) [reply lib/galaxy/tool_util/deps/mulled/mulled_build.py]: It won't just change the image name, it will not install the correct build.

Comment (bernt-matthias) [reply lib/galaxy/tool_util/deps/mulled/mulled_build.py]: Now I got it. `requirements_to_mulled_targets` also affects the names that the mulled resolvers expect: https://github.com/galaxyproject/galaxy/blob/a7fb5711431df5995b65330b5c99ac0503e8eec9/lib/galaxy/tool_util/deps/container_resolvers/mulled.py#L820C12-L820C42

I will try to finish this next week.

Sorry for the noise and thanks for double/triple checking.

File: lib/galaxy/tool_util/deps/mulled/util.py [hunk 3]
```diff
@@ -335,7 +335,14 @@ def v2_image_name(
     >>> multi_targets_versionless = [build_target("samtools"), build_target("bwa")]
     >>> v2_image_name(multi_targets_versionless)
     'mulled-v2-fe8faa35dbf6dc65a0f7f5d4ea12e31a79f73e40'
+    >>> targets_version_with_build = [build_target("samtools", version="1.3.1=h9071d68_10"), build_target("bedtools", version="2.26.0=0")]
+    >>> v2_image_name(targets_version_with_build)
+    'mulled-v2-8186960447c5cb2faa697666dc1e6d919ad23f3e:8e86df67d257ce6494ae12b2c60e1b94025ea529'
```
Comment (bernt-matthias) [inline lib/galaxy/tool_util/deps/mulled/util.py]:
```suggestion
```

File: lib/galaxy/tool_util/deps/mulled/mulled_build_tool.py [hunk 4]
```diff
@@ -20,13 +20,31 @@
...
     from galaxy.tool_util.deps.conda_util import CondaTarget
 
 
+def _mulled_build_tool(tool, args):
+    """
+    test verifies that the v2 hashes are the identical with and without using a build in the versions
+    >>> import doctest
+    >>> doctest.ELLIPSIS_MARKER = '-ignore-'
+    >>> import argparse
+    >>> _mulled_build_tool("test/functional/tools/mulled_example_multi_1.xml", argparse.Namespace(dry_run=True, base_image="does-not-matter-here-but-test-is-fast", command="build", verbose=True, involucro_path="./involucro"))  # doctest: +ELLIPSIS
```
Comment (mvdbeek) [inline lib/galaxy/tool_util/deps/mulled/mulled_build_tool.py]: This shouldn't be a doctest, and we don't want to run this with the general unit tests, just the mulled unit tests please.

**Review by mvdbeek — APPROVED** [review]
Thanks, I think this looks good to me now!

File: lib/galaxy/tool_util/deps/mulled/mulled_build_tool.py [hunk 5]
```diff
@@ -20,13 +20,31 @@
...
+    -ignore- TARGETS=samtools=1.3.1,bedtools=2.26.0 -ignore- REPO=quay.io/biocontainers/mulled-v2-8186960447c5cb2faa697666dc1e6d919ad23f3e:a6419f25efff953fc505dbd5ee734856180bb619-0 -ignore-
+    >>> _mulled_build_tool("test/functional/tools/mulled_example_multi_2.xml", argparse.Namespace(dry_run=True, base_image="does-not-matter-here-but-test-is-fast", command="build", verbose=True, involucro_path="./involucro"))  # doctest: +ELLIPSIS
+    -ignore- TARGETS=samtools=1.3.1=h9071d68_10,bedtools=2.26.0=0 -ignore- REPO=quay.io/biocontainers/mulled-v2-8186960447c5cb2faa697666dc1e6d919ad23f3e:a6419f25efff953fc505dbd5ee734856180bb619-0 -ignore-
+    """
```
Comment (bernt-matthias) [inline lib/galaxy/tool_util/deps/mulled/mulled_build_tool.py]:
```suggestion
```

**Comment by bernt-matthias** [comment]
Anything else to do here?

File: test/integration/test_container_resolvers.py [hunk 6]
```diff
@@ -480,6 +480,11 @@ class MulledTestCase:
     mulled_hash = "mulled-v2-8186960447c5cb2faa697666dc1e6d919ad23f3e:a6419f25efff953fc505dbd5ee734856180bb619-0"
 
 
+class MulledTestCaseWithBuildInfo:
+    tool_id = "mulled_example_multi_2"
```
Comment (mvdbeek) [inline test/integration/test_container_resolvers.py:484]: How does this work if you don't add the tool to the sample tool config file ?

Comment (bernt-matthias) [reply test/integration/test_container_resolvers.py:484]: Good point. I forgot to add an actual test case.

File: test/functional/tools/mulled_example_multi_2.xml (hunk 1 above)
Comment (bernt-matthias) [reply test/functional/tools/mulled_example_multi_2.xml]: Removed the test.

**Comment by mvdbeek** [comment]
If github let's me I will merge this 😆

**Pull request merged**